extensions/standard-processors/processors/TailFile.cpp

```diff
@@ -403,102 +403,141 @@ void TailFile::onTrigger(const std::shared_ptr<core::ProcessContext> &context, c
-          storeState(context);
-        }
+    did_something |= processSingleFile(context, session, fileName, state);
+    return did_something;
+  }
+
+  bool TailFile::processSingleFile(const std::shared_ptr<core::ProcessContext> &context,
+                                   const std::shared_ptr<core::ProcessSession> &session,
+                                   const std::string &fileName,
+                                   TailState &state) {
+    std::string full_file_name = state.fileNameWithPath();
+
+    if (utils::file::FileUtils::file_size(full_file_name) == 0u) {
+      logger_->log_warn("Unable to read file %s as it does not exist or has size zero", full_file_name);
+      return false;
+    }
+    logger_->log_debug("Tailing file %s from %llu", full_file_name, state.position_);
+
+  std::size_t last_dot_position = fileName.find_last_of('.');
+  std::string baseName = fileName.substr(0, last_dot_position);
+  std::string extension = fileName.substr(last_dot_position + 1);
+
+  if (!delimiter_.empty()) {
+    char delim = delimiter_[0];
+    logger_->log_trace("Looking for delimiter 0x%X", delim);
+
+    std::vector<std::shared_ptr<FlowFileRecord>> flowFiles;
+    uint64_t checksum = state.checksum_;
+    session->import(full_file_name, flowFiles, state.position_, delim, &checksum);
+    logger_->log_info("%u flowfiles were received from TailFile input", flowFiles.size());
+
+    for (auto &ffr : flowFiles) {
+      logger_->log_info("TailFile %s for %u bytes", fileName, ffr->getSize());
+      std::string logName = baseName + "." + std::to_string(state.position_) + "-" +
+                            std::to_string(state.position_ + ffr->getSize()) + "." + extension;
+      ffr->updateKeyedAttribute(PATH, state.path_);
+      ffr->addKeyedAttribute(ABSOLUTE_PATH, full_file_name);
+      ffr->updateKeyedAttribute(FILENAME, logName);
+      session->transfer(ffr, Success);
+      state.position_ += ffr->getSize() + 1;
+      state.timestamp_ = getTimeMillis();
+      state.checksum_ = checksum;
+      storeState(context);
+    }
+
+    if (!flowFiles.empty()) {
+      return true;
+    }
+  } else {
+    std::shared_ptr<FlowFileRecord> flowFile = std::static_pointer_cast<FlowFileRecord>(session->create());
+    if (flowFile) {
+      flowFile->updateKeyedAttribute(PATH, state.path_);
+      flowFile->addKeyedAttribute(ABSOLUTE_PATH, full_file_name);
+      uint64_t checksum = state.checksum_;
+      session->import(full_file_name, flowFile, true, state.position_, &checksum);
+      session->transfer(flowFile, Success);
+      logger_->log_info("TailFile %s for %llu bytes", fileName, flowFile->getSize());
+      std::string logName = baseName + "." + std::to_string(state.position_) + "-" +
+                            std::to_string(state.position_ + flowFile->getSize()) + "."
+                            + extension;
+      flowFile->updateKeyedAttribute(FILENAME, logName);
+      state.position_ += flowFile->getSize();
```

Comment:
Wonder if this could share some code with the loop above if we introduce a function. 
Just an idea, it might make it even worse :)

Reply:
I think we should get rid of the empty delimiter case (which puts everything in the log file since the last read into a single flow file, and doesn't seem useful, nor is it supported in NiFi) at some point, probably in a major release.

Until then, I would keep it like this, both to remind ourselves that this needs to be removed, and to make the removal easier.